File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return validate(root,LONG_MIN,LONG_MAX); 
    }
    bool validate(TreeNode* root, long min_val, long max_val){
        if(!root) return true; 

        if(root->val >= max_val || root->val <= min_val){
            return false; 
        }

        return validate(root->left, min_val, root->val) && validate(root->right, root->val, max_val); 
    }
};

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        if(root.left == null && root.right == null) return true; 
        return isValidBSTHelper(root, Long.MAX_VALUE, Long.MIN_VALUE); 
    }

    public boolean isValidBSTHelper(TreeNode root, long min_value, long max_value){
        if(root == null) return true; 

        if(root.val >= min_value || root.val <= max_value){
            return false; 
        }

        boolean left = isValidBSTHelper(root.left, Math.min(min_value, root.val), max_value); 
        boolean right = isValidBSTHelper(root.right, min_value, Math.max(max_value, root.val)); 

        return left && right; 
    }
}